core/src/trezor/ui/text.py

```diff
@@ -15,109 +15,255 @@
 
-        if isinstance(word, int):
-            if word is BR or word is BR_HALF:
+        if isinstance(item, int):
+            if item is BR or item is BR_HALF:
                 # line break or half-line break
                 if offset_y > offset_y_max:
                     ui.display.text(offset_x, offset_y, "...", ui.BOLD, ui.GREY, bg)
                     return
                 offset_x = INITIAL_OFFSET_X
-                offset_y += TEXT_LINE_HEIGHT if word is BR else TEXT_LINE_HEIGHT_HALF
-            elif word in FONTS:
+                offset_y += TEXT_LINE_HEIGHT if item is BR else TEXT_LINE_HEIGHT_HALF
+            elif item in _FONTS:
                 # change of font style
-                font = word
+                font = item
+                SPACE = ui.display.text_width(" ", font)
             else:
                 # change of foreground color
-                fg = word
+                fg = item
             continue
 
-        width = ui.display.text_width(word, font)
-
-        while offset_x + width > offset_x_max or (
-            has_next_word and offset_y >= offset_y_max
+        # XXX hack:
+        # if the upcoming word does not fit on this line but fits on the following,
+        # render it after a linebreak
+        item_width = ui.display.text_width(item, font)
+        if (
+            item_width <= line_width
+            and item_width + offset_x - INITIAL_OFFSET_X > line_width
         ):
-            beginning_of_line = offset_x == INITIAL_OFFSET_X
-            word_fits_in_one_line = width < (offset_x_max - INITIAL_OFFSET_X)
-            if (
-                offset_y < offset_y_max
-                and word_fits_in_one_line
-                and not beginning_of_line
-            ):
-                # line break
-                offset_x = INITIAL_OFFSET_X
-                offset_y += TEXT_LINE_HEIGHT
-                break
-            # word split
-            if offset_y < offset_y_max:
-                split = "-"
-                splitw = DASH
-            else:
-                split = "..."
-                splitw = ELLIPSIS
-            # find span that fits
-            for index in range(len(word) - 1, 0, -1):
-                letter = word[index]
-                width -= ui.display.text_width(letter, font)
-                if offset_x + width + splitw < offset_x_max:
-                    break
-            else:
-                index = 0
-            span = word[:index]
-            # render word span
-            ui.display.text(offset_x, offset_y, span, font, fg, bg)
-            ui.display.text(offset_x + width, offset_y, split, ui.BOLD, ui.GREY, bg)
-            # line break
-            if offset_y >= offset_y_max:
+            offset_y += TEXT_LINE_HEIGHT
+            ui.display.text(INITIAL_OFFSET_X, offset_y, item, font, fg, bg)
+            offset_x = INITIAL_OFFSET_X + item_width + SPACE
+            continue
+
+        span.reset(
+            item,
+            char_offset,
```

Comment:
Shouldn't this apply only to the first item? I.e. something like `char_offset if item_index == item_offset else 0`?

Reply:
fixed in 0e13f417